Add namespace status conditions

wozniakjan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind feature
/sig api-machinery

**What this PR does / why we need it**:
For users to understand what is preventing the namespace deletion and not remove `kubernetes` finalizer from namespaces: https://github.com/kubernetes/kubernetes/issues/64002, https://github.com/kubernetes/kubernetes/issues/60807, https://github.com/kubernetes/kubernetes/issues/66735,
https://github.com/kubernetes/kubernetes/issues/72244

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #70916

**Special notes for your reviewer**:
This is a parallel effort to https://github.com/kubernetes/kubernetes/pull/71457, the issue #70916 discusses both implementations.

Sample output
```yaml
$ kubectl get -o yaml ns test
apiVersion: v1
kind: Namespace
metadata:
  creationTimestamp: "2019-02-13T12:54:40Z"
  deletionTimestamp: "2019-02-13T12:55:11Z"
  name: test
  resourceVersion: "1150486"
  selfLink: /api/v1/namespaces/test
  uid: 870e9db8-2f8e-11e9-9c34-08002711add2
spec:
  finalizers:
  - kubernetes
status:
  conditions:
  - lastTransitionTime: "2019-02-13T12:58:03Z"
    message: All content successfully deleted
    reason: ContentDeleted
    status: "False"
    type: NamespaceDeletionContentFailure
  - lastTransitionTime: "2019-02-13T12:55:16Z"
    message: 'Discovery failed for some groups, 2 failing: unable to retrieve the
      complete list of server APIs: mutators.kubedb.com/v1alpha1: the server is currently
      unable to handle the request, validators.kubedb.com/v1alpha1: the server is
      currently unable to handle the request'
    reason: DiscoveryFailed
    status: "True"
    type: NamespaceDeletionDiscoveryFailure
  phase: Terminating
```

**Does this PR introduce a user-facing change?**:
```release-note
Add status condition to namespace resource
```

TODO:
- [x] unit tests
- [ ] ~e2e tests~
- [x] docs - https://github.com/kubernetes/community/blob/master/contributors/design-proposals/architecture/namespaces.md - https://github.com/kubernetes/community/pull/4045
- [x] kube-state-metrics - https://github.com/kubernetes/kube-state-metrics/blob/d96e9a8ad1b1b4de4d996b4d81b6f4bc693fe67f/internal/collector/namespace.go - https://github.com/kubernetes/kube-state-metrics/pull/949

derekwaynecarr commented:
i agree with improving transparency on why a namespace is stuck terminating.

can we ensure the design documentation for namespaces are kept in sync:
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/architecture/namespaces.md

in practice it would be good to know a listing of what resources are not fetchable so an admin could choose to manually remove the finalizer safely if required.  its not clear from the existing pr if you had multiple aggregated APIs not responding, the condition gives you a safe enough response.

sttts commented on pkg/apis/core/types.go:
```diff
@@ -3785,6 +3787,30 @@ const (
+	NamespaceDeletionEstimate NamespaceConditionType = "DeletionEstimate"
+)
+
+// Namespace Condition contains details about state of namespace
```
nit: without the space between Namespace and Condition

sttts commented on pkg/controller/namespace/deletion/namespaced_resources_deleter.go:
```diff
@@ -563,3 +567,64 @@ func (d *namespacedResourcesDeleter) estimateGracefulTerminationForPods(ns strin
+			return err
+		}
+		changeStatusConditionWithErr(ns, deletionErr)
+		_, err = d.nsClient.UpdateStatus(ns)
```
this updates always due to the probe time. This deserves a comment here why we don't call DeepEqual on old and new status.

wozniakjan replied:
thanks, I absolutely missed that. I would like to use your comment and check status update before sending to the API server

sttts commented on pkg/controller/namespace/deletion/namespaced_resources_deleter.go:
```diff
@@ -135,10 +136,13 @@ func (d *namespacedResourcesDeleter) Delete(nsName string) error {
 	// there may still be content for us to remove
 	estimate, err := d.deleteAllContent(namespace.Name, *namespace.DeletionTimestamp)
 	if err != nil {
+		d.updateStatusCondition(namespace.Name, err)
```
how often does this happen? Seconds? Milliseconds? With exponential backoff?

lavalamp replied:
You need a code path that will remove the error condition if it eventually successfully deletes the thing, I think?

wozniakjan replied:
at the moment, if the status condition no longer applies then the namespace is deleted. But I would like to re-think my proposed implementation based on your comments and after that a delete path will be probably needed

lavalamp commented on pkg/controller/namespace/deletion/namespaced_resources_deleter.go:
```diff
@@ -135,10 +136,13 @@ func (d *namespacedResourcesDeleter) Delete(nsName string) error {
 	}
 	if estimate > 0 {
-		return &ResourcesRemainingError{estimate}
+		est := &ResourcesRemainingError{estimate}
```
Instead of making two calls, maybe consider making a list of conditions that you want to set, and then when you get here, make a single call that sets them all at once?

lavalamp commented on pkg/controller/namespace/deletion/namespaced_resources_deleter.go:
```diff
@@ -563,3 +567,69 @@ func (d *namespacedResourcesDeleter) estimateGracefulTerminationForPods(ns strin
+			fmt.Sprintf("There are remaining resources, pending deletion estimate %vs", estimate.Estimate))
+	} else {
+		return changeStatusCondition(ns, v1.NamespaceDeletionFailure, "DeletionFailed",
+			fmt.Sprintf("There are remaining resources, deletion error: %v", err))
```
I recommend not having this function and figuring out the conditions you want before even calling `updateStatusCondition`. For example, if 3 things can't be deleted, I think it'd be good if the error message called out the number of things that can't be deleted. It should also probably mention e.g. the alphabetically first one.

I think the code as written will have a lot of churn, randomly switching the message if multiple things can't be deleted?

wozniakjan replied:
after doing a little bit more testing, that is exactly what is happening. Alphabetical order sounds like a good solution, thanks!

lavalamp commented on pkg/controller/namespace/deletion/namespaced_resources_deleter.go:
```diff
@@ -563,3 +567,69 @@ func (d *namespacedResourcesDeleter) estimateGracefulTerminationForPods(ns strin
+// does not propagate its error further, instead it logs it and the error
+// from namespace deletion is propagated.
+func (d *namespacedResourcesDeleter) updateStatusCondition(nsName string, deletionErr error) {
+	err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
```
I recommend not retrying--if it fails, just wait until the whole thing gets called again?

lavalamp reviewed (commented):
Some initial comments

I'm really excited about getting visibility into the namespace lifecycle controller!

wozniakjan commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4245,6 +4251,32 @@ const (
+	// NamespaceDeletionFailure contains information about namespace deleter errors
+	NamespaceDeletionFailure NamespaceConditionType = "DeletionFailure"
+	// NamespaceDeletionEstimate contains information about successful namespace deletion operation with pending estimate
+	NamespaceDeletionEstimate NamespaceConditionType = "DeletionEstimate"
```
that is good food for thought, I am starting to see the benefits of doing a bit more than just passing the error from deleter to a namespace condition. And I think this is also what @derekwaynecarr was trying to explain above. I will try to rethink this over the weekend.

wozniakjan commented:
@lavalamp I pushed some code where I was trying to address initial round of comments, quick digest of the conversation above

> Is a condition really the best way of phrasing an estimate?

probably not, I dropped it from the status conditions. I think the estimate is only for pods 
https://github.com/kubernetes/kubernetes/blob/40c91a0951de782a0e3152a3998d6c2259328b83/pkg/controller/namespace/deletion/namespaced_resources_deleter.go#L522-L525
and the default graceful termination should be 30s, so it could be acceptable to omit this feedback to an end user for now

> I recommend not having this function and figuring out the conditions you want before even calling `updateStatusCondition`

I like like your recommendation and I tried to implement this in newly added `NamespaceConditionUpdater`. This should be responsible for handling namespace conditions related to namespace deletion (currently there are no other conditions)

> it'd be good if the error message called out the number of things that can't be deleted. It should also probably mention e.g. the alphabetically first one.

that sounds very reasonable, added that. I also split the condition into three based on the deleter error types to improve UX if multiple things failed.
<details>
  <summary>Old example of status conditions</summary>
  
```yaml
apiVersion: v1
kind: Namespace
metadata:
  creationTimestamp: "2019-02-12T08:49:30Z"
  deletionTimestamp: "2019-02-12T08:51:05Z"
  name: test3
  resourceVersion: "1045153"
  selfLink: /api/v1/namespaces/test3
  uid: 1cc27c77-2ea3-11e9-8c82-08002711add2
spec:
  finalizers:
  - kubernetes
status:
  conditions:
  - lastProbeTime: "2019-02-12T12:37:47Z"
    lastTransitionTime: "2019-02-12T08:51:14Z"
    message: 'There are remaining resources, deletion error: [unable to retrieve the
      complete list of server APIs: mutators.kubedb.com/v1alpha1: the server is currently
      unable to handle the request, validators.kubedb.com/v1alpha1: the server is
      currently unable to handle the request, Internal error occurred: failed calling
      webhook "mysql.validators.kubedb.com": the server is currently unable to handle
      the request, Internal error occurred: failed calling webhook "redis.validators.kubedb.com":
      the server is currently unable to handle the request, Internal error occurred:
      failed calling webhook "elasticsearch.validators.kubedb.com": the server is
      currently unable to handle the request, Internal error occurred: failed calling
      webhook "dormantdatabase.validators.kubedb.com": the server is currently unable
      to handle the request, Internal error occurred: failed calling webhook "memcached.validators.kubedb.com":
      the server is currently unable to handle the request, Internal error occurred:
      failed calling webhook "mongodb.validators.kubedb.com": the server is currently
      unable to handle the request, Internal error occurred: failed calling webhook
      "postgres.validators.kubedb.com": the server is currently unable to handle the
      request, Internal error occurred: failed calling webhook "etcd.validators.kubedb.com":
      the server is currently unable to handle the request]'
    reason: DeletionFailed
    status: "True"
    type: DeletionFailure
  phase: Terminating
```
</details>

<details>
  <summary>New example of status conditions</summary>

```yaml
apiVersion: v1
kind: Namespace
metadata:
  creationTimestamp: "2019-02-13T12:54:40Z"
  deletionTimestamp: "2019-02-13T12:55:11Z"
  name: test
  resourceVersion: "1150211"
  selfLink: /api/v1/namespaces/test
  uid: 870e9db8-2f8e-11e9-9c34-08002711add2
spec:
  finalizers:
  - kubernetes
status:
  conditions:
  - lastProbeTime: "2019-02-13T12:55:16Z"
    lastTransitionTime: "2019-02-13T12:55:16Z"
    message: 'Failed to delete all resource types, 8 remaining: Internal error occurred:
      failed calling webhook "dormantdatabase.validators.kubedb.com": the server is
      currently unable to handle the request, Internal error occurred: failed calling
      webhook "elasticsearch.validators.kubedb.com": the server is currently unable
      to handle the request, Internal error occurred: failed calling webhook "etcd.validators.kubedb.com":
      the server is currently unable to handle the request, Internal error occurred:
      failed calling webhook "memcached.validators.kubedb.com": the server is currently
      unable to handle the request, Internal error occurred: failed calling webhook
      "mongodb.validators.kubedb.com": the server is currently unable to handle the
      request, Internal error occurred: failed calling webhook "mysql.validators.kubedb.com":
      the server is currently unable to handle the request, Internal error occurred:
      failed calling webhook "postgres.validators.kubedb.com": the server is currently
      unable to handle the request, Internal error occurred: failed calling webhook
      "redis.validators.kubedb.com": the server is currently unable to handle the
      request'
    reason: ContentDeletionFailed
    status: "True"
    type: NamespaceDeletionContentFailure
  - lastProbeTime: "2019-02-13T12:55:16Z"
    lastTransitionTime: "2019-02-13T12:55:16Z"
    message: 'Discovery failed for some groups, 2 failing: unable to retrieve the
      complete list of server APIs: mutators.kubedb.com/v1alpha1: the server is currently
      unable to handle the request, validators.kubedb.com/v1alpha1: the server is
      currently unable to handle the request'
    reason: DiscoveryFailed
    status: "True"
    type: NamespaceDeletionDiscoveryFailure
  phase: Terminatin
```
</details>
_

> You need a code path that will remove the error condition if it eventually successfully deletes the thing, I think?

now, absolutely. I looked at other controllers and they seem to invalidate the condition by setting `Type` to `v1.ConditionTypeFalse` and setting appropriate message, reason, transition timestamp. So that is what I ended up doing

<details>
  <summary>Example of updated condition</summary>

```yaml
apiVersion: v1
kind: Namespace
metadata:
  creationTimestamp: "2019-02-13T12:54:40Z"
  deletionTimestamp: "2019-02-13T12:55:11Z"
  name: test
  resourceVersion: "1150486"
  selfLink: /api/v1/namespaces/test
  uid: 870e9db8-2f8e-11e9-9c34-08002711add2
spec:
  finalizers:
  - kubernetes
status:
  conditions:
  - lastProbeTime: "2019-02-13T12:55:16Z"
    lastTransitionTime: "2019-02-13T12:58:03Z"
    message: All content successfully deleted
    reason: ContentDeleted
    status: "False"
    type: NamespaceDeletionContentFailure
  - lastProbeTime: "2019-02-13T12:55:16Z"
    lastTransitionTime: "2019-02-13T12:55:16Z"
    message: 'Discovery failed for some groups, 2 failing: unable to retrieve the
      complete list of server APIs: mutators.kubedb.com/v1alpha1: the server is currently
      unable to handle the request, validators.kubedb.com/v1alpha1: the server is
      currently unable to handle the request'
    reason: DiscoveryFailed
    status: "True"
    type: NamespaceDeletionDiscoveryFailure
  phase: Terminating
```
</details>
_

> this updates always due to the probe time. This deserves a comment here why we don't call DeepEqual on old and new status.

And I kept the earlier implemented condition to only update condition status if the message for particular condition changes.

Looking forward to the next round of feedback. If you find the design acceptable, I would like to continue with unit and e2e tests, and finally proposing documentation changes.

lavalamp commented:
Apologies for radio silence, I lost track of this and just realized it's my turn. @sttts or @derekwaynecarr might get to it before I have a chance though.

/ok-to-test

wozniakjan commented:
no problem and thanks for the tests. I will try to move this forward following weekend

deads2k commented on pkg/controller/namespace/deletion/namespaced_resources_deleter.go:
```diff
@@ -135,10 +136,13 @@ func (d *namespacedResourcesDeleter) Delete(nsName string) error {
 	// there may still be content for us to remove
 	estimate, err := d.deleteAllContent(namespace.Name, *namespace.DeletionTimestamp)
 	if err != nil {
+		d.updateStatusCondition(namespace.Name, err)
```
You need to clear this in cases where you had an error and then cleared that error.  Looking below, you can see that when this finishes, we issue an update, not a delete.

yastij commented:
ping @wozniakjan - is this still active work ?

/priority important-soon

wozniakjan commented:
Thanks for the ping @yastij. I have temporarily put this on hold waiting for reviews and I think it may have been a bit too long. I will put this back on my TODO list. Hope to be able to find some time this weekend.

wozniakjan committed:
Add status conditions to namespaces

wozniakjan commented:
/test pull-kubernetes-e2e-gce-100-performance

yastij commented:
@wozniakjan - is this ready for another round of reviews ?

/assign

wozniakjan commented:
reviews, yes please. The rebased and pushed implements - https://github.com/kubernetes/kubernetes/pull/73405#issuecomment-463190495, still a little unclear for me whether this is an acceptable approach or not

I will work on adding unit tests

wozniakjan commented:
@yastij adding a simple unit test covering this part and checking update on namespace status, I will be looking at e2e tests to see if there is any new needed. Any reviews/comments are appreciated.

wozniakjan commented:
/test pull-kubernetes-integration

wozniakjan commented:
/test pull-kubernetes-integration

wozniakjan committed:
Add tests for namespace status conditions

yastij commented:
@wozniakjan - can you rebase this please ?

yastij commented:
@kubernetes/api-approvers - given that this adds new fields in a stable API, I guess it should have a KEP ?

wozniakjan commented:
@yastij thanks for pinging, rebased

tariq1890 commented on pkg/controller/namespace/deletion/status_condition_utils.go:
```diff
@@ -0,0 +1,171 @@
+/*
+Copyright 2015 The Kubernetes Authors.
```
```suggestion
Copyright 2019 The Kubernetes Authors.
```

tariq1890 commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4340,6 +4346,32 @@ const (
+	NamespaceDeletionGVParsingFailure NamespaceConditionType = "NamespaceDeletionGroupVersionParsingFailure"
+)
+
+// NamespaceCondition contains details about state of namespace
```
```suggestion
// NamespaceCondition contains details about state of namespace.
```

tariq1890 commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4340,6 +4346,32 @@ const (
+	NamespaceDeletionDiscoveryFailure NamespaceConditionType = "NamespaceDeletionDiscoveryFailure"
+	// NamespaceDeletionContentFailure contains information about namespace deleter errors during deletion of resources
+	NamespaceDeletionContentFailure NamespaceConditionType = "NamespaceDeletionContentFailure"
+	// NamespaceDeletionGVParsingFailure contains information about namespace deleter errors parsing GV for legacy types
```
```suggestion
	// NamespaceDeletionGVParsingFailure contains information about namespace deleter errors parsing GV for legacy types.
```

tariq1890 commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4340,6 +4346,32 @@ const (
+const (
+	// NamespaceDeletionDiscoveryFailure contains information about namespace deleter errors during resource discovery
+	NamespaceDeletionDiscoveryFailure NamespaceConditionType = "NamespaceDeletionDiscoveryFailure"
+	// NamespaceDeletionContentFailure contains information about namespace deleter errors during deletion of resources
```
```suggestion
	// NamespaceDeletionContentFailure contains information about namespace deleter errors during deletion of resources.
```

tariq1890 commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4340,6 +4346,32 @@ const (
+
+// These are valid conditions of a namespace
+const (
+	// NamespaceDeletionDiscoveryFailure contains information about namespace deleter errors during resource discovery
```
```suggestion
	// NamespaceDeletionDiscoveryFailure contains information about namespace deleter errors during resource discovery.
```

tariq1890 commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4340,6 +4346,32 @@ const (
 
+type NamespaceConditionType string
+
+// These are valid conditions of a namespace
```
```suggestion
// These are valid conditions of a namespace.
```

tariq1890 commented on pkg/controller/namespace/namespace_controller.go:
```diff
@@ -145,7 +145,7 @@ func (nm *NamespaceController) worker() {
 			// rather than wait for a full resync, re-add the namespace to the queue to be processed
 			nm.queue.AddRateLimited(key)
-			utilruntime.HandleError(err)
+			utilruntime.HandleError(fmt.Errorf("Deletion of namespace %v failed: %v", key, err))
```
```suggestion
			utilruntime.HandleError(errors.Errorf("Deletion of namespace %v failed: %v", key, err))
```

tariq1890 replied:
Let's use `errors.Wrapf` here.

tariq1890 commented on pkg/controller/namespace/deletion/status_condition_utils.go:
```diff
@@ -0,0 +1,171 @@
+
+}
+
+// ProcessDeleteContentErr creates error condition from multiple delete content errors
```
```suggestion
// ProcessDeleteContentErr creates error condition from multiple delete content errors.
```

tariq1890 commented on pkg/controller/namespace/deletion/status_condition_utils.go:
```diff
@@ -0,0 +1,171 @@
+	u.deleteContentErrors = append(u.deleteContentErrors, err)
+}
+
+// Update compiles processed errors from namespace deletion into status conditions
```
```suggestion
// Update compiles processed errors from namespace deletion into status conditions.
```

tariq1890 commented on pkg/apis/core/types.go:
```diff
@@ -3867,6 +3869,29 @@ const (
+	NamespaceDeletionGVParsingFailure NamespaceConditionType = "NamespaceDeletionGroupVersionParsingFailure"
+)
+
+// NamespaceCondition contains details about state of namespace
```
```suggestion
// NamespaceCondition contains details about state of namespace.
```

tariq1890 commented on pkg/apis/core/types.go:
```diff
@@ -3867,6 +3869,29 @@ const (
 
+type NamespaceConditionType string
+
+// These are valid conditions of a namespace
```
```suggestion
// These are valid conditions of a namespace.
```

wozniakjan commented:
thanks @tariq1890 for your suggestions, included them in the PR and rebased on latest master while I was at it.

Any update with reviews @yastij ?

tariq1890 commented on pkg/apis/core/types.go:
```diff
@@ -3997,6 +3999,29 @@ const (
 	NamespaceTerminating NamespacePhase = "Terminating"
 )
 
+type NamespaceConditionType string
```
Let's add a godoc for `NamespaceConditionType`.

tariq1890 commented on pkg/controller/namespace/deletion/namespaced_resources_deleter.go:
```diff
@@ -480,35 +480,47 @@ func (d *namespacedResourcesDeleter) deleteAllContentForGroupVersionResource(
 	if len(errs) > 0 {
+		if hasChanged := conditionUpdater.Update(ns); hasChanged {
+			if _, err = d.nsClient.UpdateStatus(ns); err != nil {
+				utilruntime.HandleError(fmt.Errorf("Couldn't update status condition for namespace %q: %v", namespace, err))
```
Let's use `errors.Wrapf` here.

wozniakjan replied:
Given this package uses `fmt.Errorf()` multiple times and does not use `errors.Wrapf()` or `errors.Wrap()`, it may be better to stay in sync with the package conventions, what do you think?

Also `pkg/controller/` seem to be more old-school :)
```
$ grep pkg/controller/ -nrie Wrapf --include=*.go | wc -l
0
$ grep pkg/controller/ -nrie Errorf --include=*.go | wc -l
1771
```

tariq1890 replied:
Makes sense.

tariq1890 commented on pkg/controller/namespace/deletion/status_condition_utils.go:
```diff
@@ -0,0 +1,171 @@
+	}
+)
+
+// ProcessGroupVersionErr creates error condition if parsing GroupVersion of resources fails
```
```suggestion
// ProcessGroupVersionErr creates error condition if parsing GroupVersion of resources fails.
```

tariq1890 commented on pkg/controller/namespace/deletion/status_condition_utils.go:
```diff
@@ -0,0 +1,171 @@
+	u.newConditions = append(u.newConditions, d)
+}
+
+// ProcessDiscoverResourcesErr creates error condition from ErrGroupDiscoveryFailed
```
```suggestion
// ProcessDiscoverResourcesErr creates error condition from ErrGroupDiscoveryFailed.
```

tariq1890 commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4608,6 +4614,32 @@ const (
 	NamespaceTerminating NamespacePhase = "Terminating"
 )
 
+type NamespaceConditionType string
```
Let's add  godoc here too.

tariq1890 reviewed (commented):
@wozniakjan Just added some more comments. PR looks really good. Note that I am not a maintainer of `kubernetes/kubernetes`. Just helping out :).

tariq1890 commented:
Pinging @thockin for attention on this.

wozniakjan commented:
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-integration

wozniakjan commented on pkg/controller/namespace/deletion/namespaced_resources_deleter.go:
```diff
@@ -480,35 +480,47 @@ func (d *namespacedResourcesDeleter) deleteAllContentForGroupVersionResource(
 	if len(errs) > 0 {
+		if hasChanged := conditionUpdater.Update(ns); hasChanged {
+			if _, err = d.nsClient.UpdateStatus(ns); err != nil {
+				utilruntime.HandleError(fmt.Errorf("Couldn't update status condition for namespace %q: %v", namespace, err))
```
good catch, made it lowercase

wozniakjan commented:
/test pull-kubernetes-integration

yastij reviewed (commented):
/lgtm
/retest

yastij commented:
/retest

wozniakjan commented:
> Sorry its not clear from the sample output which objects in the namespace is preventing the deletion of the namespace or is that just a positive case. 

in the sample output, it is actually the discovery that was failing. Discovery phase is a prerequisite to the phase where deleter starts deleting objects. The deleter algorithm is written in best-effort kind of way where it tries to delete whatever it can and then reports all the errors it encountered along the way.

> The way I understand this PR is that it injects additional conditions that will tell us what is preventing the deletion ?

```yaml
  - lastTransitionTime: "2019-02-13T12:55:16Z"
    message: 'Discovery failed for some groups, 2 failing: unable to retrieve the
      complete list of server APIs: mutators.kubedb.com/v1alpha1: the server is currently
      unable to handle the request, validators.kubedb.com/v1alpha1: the server is
      currently unable to handle the request'
    reason: DiscoveryFailed
    status: "True"
    type: NamespaceDeletionDiscoveryFailure
```

this would be the status condition telling you that discovery failed for 2 groups - `mutators.kubedb.com/v1alpha1` and `validators.kubedb.com/v1alpha1`. Does this help?

wozniakjan commented:
/test pull-kubernetes-bazel-test

xmudrii commented:
As this PR is targeting 1.16, I'm adding it to the milestone
/milestone v1.16

krmayankk commented:
> > Sorry its not clear from the sample output which objects in the namespace is preventing the deletion of the namespace or is that just a positive case.
> 
> in the sample output, it is actually the discovery that was failing. Discovery phase is a prerequisite to the phase where deleter starts deleting objects. The deleter algorithm is written in best-effort kind of way where it tries to delete whatever it can and then reports all the errors it encountered along the way.
> 
> > The way I understand this PR is that it injects additional conditions that will tell us what is preventing the deletion ?
> 
> ```yaml
>   - lastTransitionTime: "2019-02-13T12:55:16Z"
>     message: 'Discovery failed for some groups, 2 failing: unable to retrieve the
>       complete list of server APIs: mutators.kubedb.com/v1alpha1: the server is currently
>       unable to handle the request, validators.kubedb.com/v1alpha1: the server is
>       currently unable to handle the request'
>     reason: DiscoveryFailed
>     status: "True"
>     type: NamespaceDeletionDiscoveryFailure
> ```
> 
> this would be the status condition telling you that discovery failed for 2 groups - `mutators.kubedb.com/v1alpha1` and `validators.kubedb.com/v1alpha1`. Does this help?

thanks @wozniakjan In the doc bug, it would help to actually show the deletion failure case .thanks for fixing this

xmudrii commented:
@wozniakjan It seems like this PR needs to be rebased before it can be merged.

wozniakjan committed:
Update generated files

make generated_files UPDATE_API_KNOWN_VIOLATIONS=true
./hack/update-generated-protobuf.sh
./hack/update-openapi-spec.sh
./hack/update-bazel.sh
./hack/update-generated-swagger-docs.sh
./hack/update-generated-api-compatibility-data.sh

wozniakjan commented:
thanks @xmudrii, rebasing on the latest master to resolve conflicts with https://github.com/kubernetes/kubernetes/pull/79386 in staging/src/k8s.io/api/core/v1/generated.pb.go and updating generated files

deads2k commented:
/lgtm

wozniakjan commented:
/test pull-kubernetes-integration

wozniakjan commented:
/test pull-kubernetes-bazel-test

wozniakjan commented:
/test pull-kubernetes-e2e-gce-100-performance

xmudrii commented:
/retest

lavalamp commented:
/test pull-kubernetes-bazel-build
/test pull-kubernetes-bazel-test

tariq1890 commented:
🎉 🎉

liggitt commented:
bazel failure was a flake: https://github.com/kubernetes/kubernetes/issues/82150

wozniakjan commented:
Thank you everyone for reviews, suggestions, improvements and getting this in.
I will try to produce matching docs change over the weekend and then look at the changes for kube-state-metrics
